src/server/pfs/fuse/loopback.go

```diff
@@ -665,12 +666,41 @@ func (n *loopbackNode) download(ctx context.Context, origPath string, state file
 		}
 		return nil
 	}
-	projectName := ro.File.Commit.Branch.Repo.Project.GetName()
-	repoName := ro.File.Commit.Branch.Repo.Name
-	filePath := filepath.Join(parts[1:]...)
-	if err := n.c().ListFile(client.NewCommit(projectName, repoName, branch, commit), filePath, createFile); err != nil && !errutil.IsNotFoundError(err) &&
-		!pfsserver.IsOutputCommitNotFinishedErr(err) {
-		return err
+	// Calling ListFile on a repo with many files at the top level when we only care about a single
+	// file, for example, is very expensive. Hence the below optimizations.
+	// We only want to create descendants and ancestors of mounted file(s) ro.Files
+	// ro.Files[i].Path is the path of a mounted file, filePath is OS requested path
+	// Example: ro.Files[i].Path = /file, filePath = /files <--- DON'T CREATE
+	// Example: ro.Files[i].Path = /file, filePath = /file/path <--- CREATE
+	// Example: ro.Files[i].Path = /dir/file, filePath = /dir <--- CREATE
+	for _, f := range ro.Files {
+		ancestor, intermediates := isGrandparentOf(filePath, f.Path)
+		switch {
+		case ancestor:
+			path := filepath.Join(n.namePath(name), filePath, intermediates)
+			if err := errors.EnsureStack(os.MkdirAll(path, 0777)); err != nil {
+				return err
+			}
+		case isParentOf(filePath, f.Path):
+			fi, err := n.c().InspectFile(commit, f.Path)
+			if err != nil {
+				return err
+			}
+			if fi.FileType == pfs.FileType_DIR {
+				if err := errors.EnsureStack(os.MkdirAll(n.filePath(name, fi), 0777)); err != nil {
+					return err
+				}
+			} else {
+				if err := n.createFiles(commit, f.Path, createFile); err != nil {
+					return err
+				}
+			}
+		case isDescendantOf(filePath, f.Path):
+			if err := n.createFiles(commit, filePath, createFile); err != nil {
+				return err
+			}
+		default:
```

Comment:
could be good to add an assert(false) here since we expect nothing to fall through

Reply:
I'll add a comment to the code but execution CAN hit the default case. If a user creates folders in the mount, then they will enter a path that is no longer one of the previous cases. In this case, we want to ignore, since there's nothing to mount.

The mount server supports writing data to Pachyderm by creating folders and files in the mount. However, this functionality isn't exposed in the notebooks extension.